server/plugin/user_test.go

```diff
@@ -0,0 +1,223 @@
+				s.On("StoreOAuth2State", mock.AnythingOfType("string")).Return(nil)
+			},
+		},
+		{
+			description: "Error occurred while storing oauth2 state",
+			setupStore: func(s *mock_plugin.Store) {
+				s.On("LoadUser", testutils.GetID()).Return(nil, fmt.Errorf("mockErrMessage"))
+				s.On("StoreOAuth2State", mock.AnythingOfType("string")).Return(fmt.Errorf("mockErrMessage"))
+			},
+			expectedErrorMessage: "mockErrMessage",
+		},
+	} {
+		t.Run(test.description, func(t *testing.T) {
+			p := Plugin{}
+			store := mock_plugin.NewStore(t)
+
+			test.setupStore(store)
+			p.store = store
+
+			res, err := p.InitOAuth2(testutils.GetID())
+			if test.expectedErrorMessage != "" {
+				require.Equal(t, "", res)
+				require.NotNil(t, err)
+				require.Equal(t, test.expectedErrorMessage, err.Error())
+			} else {
+				require.Nil(t, err)
+				require.NotEqual(t, "", res)
+			}
+		})
+	}
+}
+
+func TestCompleteOAuth2(t *testing.T) {
+	for name, test := range map[string]struct {
+		authedUserID         string
+		code                 string
+		state                string
+		setupStore           func(*mock_plugin.Store)
+		setupAPI             func(*plugintest.API)
+		setupPlugin          func(*Plugin)
+		expectedErrorMessage string
+	}{
+		"success": {
+			authedUserID: "mockUserID",
+			code:         "mockCode",
+			state:        "mockState_mockUserID",
+			setupStore: func(s *mock_plugin.Store) {
+				s.On("VerifyOAuth2State", "mockState_mockUserID").Return(nil)
+				s.On("StoreUser", mock.AnythingOfType("*serializer.User")).Return(nil)
+			},
+			setupAPI: func(a *plugintest.API) {
+				a.On("GetUser", "mockUserID").Return(&model.User{}, nil)
+			},
+			setupPlugin: func(p *Plugin) {
+				monkey.PatchInstanceMethod(reflect.TypeOf(&oauth2.Config{}), "Exchange", func(_ *oauth2.Config, _ context.Context, _ string, _ ...oauth2.AuthCodeOption) (*oauth2.Token, error) {
+					return &oauth2.Token{}, nil
+				})
+				monkey.PatchInstanceMethod(reflect.TypeOf(p), "NewEncodedAuthToken", func(_ *Plugin, _ *oauth2.Token) (string, error) {
+					return "mockToken", nil
+				})
+				monkey.PatchInstanceMethod(reflect.TypeOf(p), "DM", func(_ *Plugin, _, _ string, _ ...interface{}) (string, error) {
+					return "", nil
+				})
+			},
+		},
+		"missing userID, code or state": {
+			authedUserID:         "",
+			setupStore:           func(s *mock_plugin.Store) {},
+			setupAPI:             func(a *plugintest.API) {},
+			setupPlugin:          func(p *Plugin) {},
+			expectedErrorMessage: "missing user, code or state",
+		},
+		"failed to verify state": {
+			authedUserID: "mockUserID",
+			code:         "mockCode",
+			state:        "mockState_mockUserID",
+			setupStore: func(s *mock_plugin.Store) {
+				s.On("VerifyOAuth2State", "mockState_mockUserID").Return(fmt.Errorf("mockErrMessage"))
+			},
+			setupAPI:             func(a *plugintest.API) {},
+			setupPlugin:          func(p *Plugin) {},
+			expectedErrorMessage: "mockErrMessage",
+		},
+		"user ID mismatch": {
```

Comment:
Would be similar to above
```suggestion
		"failed to match user ID": {
```